### ISSUE TITLE
I am confused about gradient_scale in different context. 

### ISSUE THREAD

**Issue by horserma** [dmlc/mxnet#2994]
Hi, 
I am pretty confused about this gradient scale parameter. In optimizer, there is such a parameter, and in softmaxoutput symbol, there is also such a parameter, I am wondering what those roles are. For example, I want to implement  a very simple 2 gpu data parallel program. I have a simple network on one gpu and a copy of it on another gpu. Now I have a batch size of 8 and I divide it into 2 batches of 4 for each gpu. Then I simple bind them and feed the corresponding data to each of them and forward and backward. Now it comes to the update step. Since I have two sets of gradients for the same weight variable, what am I supposed to do? sum up theses two copies and divide it by 2? or divide it by 8? or I use 1.0/16 as a gradient scale parameter and pass it to the optimize? I hope you could clarify this a little. I would really appreciate it.
@piiswrong @pluskid @tqchen

**Comment by piiswrong** [comment]
gradient scale in softmaxoutput is for weighting loss functions if you have multiple of them.

gradient scale in optimizer is for normalizing against batch_size.

They are in effect multiplied togather in the end.

**Comment by horserma** [comment]
@piiswrong Thanks for your reply, but I still got a little bit confused because in the document, SoftmaxOutput has a "normalization" argument. It says the default is "null", which means nothing will be done. "batch" means  "op will normalize gradient by dividing batch size", while "valid" means " op will normalize gradient by divide sample not ignored". 
I have 2 questions:
1. Does this have the same effect if I set this to "null" and set an appropriate grad_scale?
2. Does "valid samples" mean all samples or only samples in a batch? For example, I have a multi-output softmax, say (2, 3, 100), which mean batch size=2, channel=3 and sample=100. If I set "batch", it divides by 2, right? If I use "valid" does it divide by 100? or (2x100)? If true, does it mean I dont need to change the gradient scale anymore and leave them to default 1.0?
I think this is really important to know if I want to make sure the training is under control. Thanks.
